File: LogClient/WebTracer.cs
using System.Reflection;
using System.Text;
using System.Text.Json;
using LogClient.Types;

namespace LogClient
{
    public sealed class WebTracer : BaseLogger, ITracer
    {
        readonly HttpClient _httpClient;

        readonly Product _currentProduct;

        readonly string _hostName;


        public WebTracer(string hostName, Product currentProduct)
        {
            _hostName = hostName;
            _httpClient = new HttpClient();
            _httpClient.Timeout = TimeSpan.FromSeconds(3);
            _httpClient.BaseAddress = new Uri(hostName);
            _currentProduct = currentProduct;
        }

        public async Task TraceAsync(string message, string user = null)
        {
            await TraceAsync(message, user, null, null);
        }

        public async Task TraceAsync(string message, string user, long? ticks, long? sessionId, string tag1 = null, string tag2 = null, string tag3 = null)
        {
            Func<Task> func = async () =>
            {
                Trace newTrace = new()
                {
                    Product = _currentProduct,
                    Message = message,
                    Username = user,
                    Ticks = ticks,
                    SessionId = sessionId,
                    Tag1 = tag1,
                    Tag2 = tag2,
                    Tag3 = tag3,
                };

                string json = JsonSerializer.Serialize(newTrace);
                var stringContent = new StringContent(json, Encoding.UTF8, "application/json");
                await _httpClient.PostAsync("/api/Trace", stringContent).ConfigureAwait(false);
            };


            // This method from the base class knows better how to execute logging.
            await PerformActionAsync(func);
        }

        public async Task<string> GenerateJavaScriptTracerObjectAsync(Product product)
        {
            var assembly = Assembly.GetExecutingAssembly();
            string javaScriptResource = "LogClient.WebTracer.js";
            string javaScriptText = String.Empty;

            using (Stream stream = assembly.GetManifestResourceStream(javaScriptResource))
            {
                using (StreamReader reader = new StreamReader(stream))
                {
                    javaScriptText = await reader.ReadToEndAsync();
                }
            }

            if (!String.IsNullOrEmpty(javaScriptText))
            {
                javaScriptText = javaScriptText.Replace("{{host_name}}", _hostName);
                javaScriptText = javaScriptText.Replace("{{prod_id}}", ((int)product).ToString());
            }

            return javaScriptText;
        }
    }
}

File: WebTracer.js
function WebTracer() {

    // private:
    const _host_name = '{{host_name}}';
    const _prod_id = {{prod_id}};

    // public:
    this.traceAdv = function (message, user, sessionId, tag2 = null, tag3 = null) {
        fetch(_host_name + '/api/Trace', {method: 'POST', headers: {
            'Content-type': 'application/json; charset=utf-8'
        },
            body: JSON.stringify({
                product: _prod_id,
                message: message,
                username: user,
                ticks: Date.now(),
                sessionId: sessionId,
                Tag1: 'JS',
                Tag2: tag2,
                Tag3: tag3,
            })
        });
    }

    this.trace = function (message, user) {
        fetch(_host_name + '/api/Trace', {method: 'POST', headers: {
            'Content-type': 'application/json; charset=utf-8'
        },
            body: JSON.stringify({
                product: _prod_id,
                message: message,
                username: user,
                ticks: null,
                sessionId: null,
                Tag1: 'JS',
                Tag2: null,
                Tag3: null,
            })
        });
    }
}

window.Tracer = new WebTracer();